### PR TITLE
Add logits_processors param to batch_generate

### PR DESCRIPTION
**Pull request by sethkimmel3**
Allows the batch_generate method to accept logits processors for structured decoding. This will be very useful in local data analysis for structured extraction and similar. For context, we're the creators of [Sutro](https://github.com/sutro-sh/sutro), and may use mlx-lm as a backend to be able to use the package locally, or release another OSS package entirely for local, mlx based data analysis.

It can be invoked with an xgrammar logits processor, like: 

```
import mlx.core as mx
import xgrammar
from xgrammar.kernels.apply_token_bitmask_mlx import apply_token_bitmask_mlx
from pydantic import BaseModel

class ExtractCompanyName(BaseModel):
    company_name: str

# Specify the checkpoint
model_name = "mlx-community/Qwen3-4B-4bit"
model, tokenizer = load(path_or_hf_repo=model_name)

inputs = ['The winners of the 2020 Apple Design Awards', 'Neeva Is Search Reimagined', "Niantic's AR Lessons: The Business Angle"]

system_prompt = f"Extract the name of the company from the following text. If there is no company name, return 'None'. Return the company name as a JSON object with the key 'company_name'."

# Apply the chat template and encode to tokens
prompts = [
    tokenizer.apply_chat_template(
        [
            {"role": "system", "content": system_prompt},
            {"role": "user", "content": p}
        ],
        add_generation_prompt=True,
    )
    for p in inputs
]

class XGrammarLogitsProcessor:
    def __init__(self, grammar: xgrammar.CompiledGrammar, max_rollback_tokens: int = 16):
        self.grammar = grammar
        self.max_rollback_tokens = max_rollback_tokens
        self.bitmask = None      # allocate lazily to match logits width

    def __call__(self, tokens: mx.array, logits: mx.array) -> mx.array:
        # Lazily size the bitmask to the logits width (model's vocab at runtime)
        vocab = int(logits.shape[-1])
        if self.bitmask is None or self.bitmask.shape[1] != vocab:
            self.bitmask = xgrammar.allocate_token_bitmask(1, vocab)

        # Fresh matcher per call (per row, per step)
        matcher = xgrammar.GrammarMatcher(self.grammar, max_rollback_tokens=self.max_rollback_tokens)

        # Replay the full history so the matcher is at the correct place
        # (tokens is 1-D mx.array of token ids)
        for t in tokens.tolist():
            matcher.accept_token(int(t))

        if matcher.is_terminated():
            return logits

        matcher.fill_next_token_bitmask(self.bitmask)
        return apply_token_bitmask_mlx(mx.array(self.bitmask.numpy()), logits, vocab)

hf_tokenizer = AutoTokenizer.from_pretrained(model_name)
config = AutoConfig.from_pretrained(model_name)
# This can be larger than tokenizer.vocab_size due to paddings
full_vocab_size = config.vocab_size
tokenizer_info = xgrammar.TokenizerInfo.from_huggingface(hf_tokenizer, vocab_size=full_vocab_size)
grammar_compiler = xgrammar.GrammarCompiler(tokenizer_info, max_threads=8)
compiled_grammar = grammar_compiler.compile_json_schema(ExtractCompanyName)

# Set `verbose=True` to see generation statistics
logits_processor = XGrammarLogitsProcessor(grammar=compiled_grammar)

result = batch_generate(model, tokenizer, prompts, verbose=True, max_tokens=512, logits_processors=[logits_processor])

# The returned result contains texts completions in the same order as prompts
for text in result.texts:
    print(text)
    print("-" * 100)
```

returns: 

```
{
  "company_name": "Apple"
}
----------------------------------------------------------------------------------------------------
{
  "company_name": "Neeva"
}
----------------------------------------------------------------------------------------------------
{
  "company_name": "Niantic"
}
----------------------------------------------------------------------------------------------------
```

cc: @Ubospica

### CODE REVIEW

**Comment by awni** [comment]
Thanks for the contribution. Closing as dup of #635
